### ISSUE TITLE
Memory leaked

### ISSUE THREAD

**Issue by tuoxie007** [mruby/mruby#3696]
I tested on iPhone Simulator. I just `puts` a simple array 1M times.

```objc
- (void)testMemoryLeak
{
    unsigned int stack_keep = 0;
    mrb_state *mrb = mrb_open();
    mrbc_context *mrb_ctx = mrbc_context_new(mrb);

    for (NSInteger i=0; i<1000*1000; i++) {
        struct mrb_parser_state *parser = mrb_parse_string(mrb, "puts [1, 2, 3, 4, 5]", mrb_ctx);
        if (parser == NULL) {
            return;
        }

        struct RProc *proc = mrb_generate_code(mrb, parser);
        if (proc == NULL) {
            return;
        }

        int gc_arena = mrb_gc_arena_save(mrb);
        mrb_context_run(mrb, proc, mrb_top_self(mrb), stack_keep);
        mrb_gc_arena_restore(mrb, gc_arena);
        stack_keep = proc->body.irep->nlocals;
        mrb_parser_free(parser);
        if (mrb->exc) {
            return;
        }
    }
}
```

Memory was continuous increasing like this. Never decreased.

<img width="408" alt="qq20170607-173455 2x" src="https://user-images.githubusercontent.com/1240920/26872156-2806db14-4ba8-11e7-9966-e623e48fcfb8.png">

I seems that some memory was leaked. Is there any way to release garbage memory?

**Comment by matz** [comment]
Try calling `mrb_close(mrb)` before exiting the function.
I will investigate further memory leak when I have spare time.

**Comment by texrg** [comment]
is possible to disabling all garbage collector functions?

**Comment by matz** [comment]
You can call `GC.disable`.
From C function, you can `mrb->gc.disabled = TRUE;`.

**Comment by texrg** [comment]
@matz thanks

Is possible add disabling GC on 
~/build_config.rb or ~/include/mrbconf.h ?

**Comment by matz** [comment]
No. Without GC, a very small program could run for short amount of time.
You may want to try `mruby/c` from http://github.com/mrubyc/mrubyc

**Comment by matz** [comment]
Closing — the leak disappears with the high-level API. `mrb_load_string_cxt` (or `mrb_load_string`) handles the irep lifecycle automatically; the low-level pipeline (`mrb_parse_string` + `mrb_generate_code` + `mrb_vm_run`) requires the caller to release the generated irep via `mrb_irep_decref(mrb, proc->body.irep)`. Verified at 1000 iterations: high-level API → 0 bytes in use at exit; low-level without decref → memory accumulates.
